### PR TITLE
KAFKA-9410: Make groupId Optional in KafkaConsumer

### PR DESCRIPTION
**Pull request by belugabehr**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by rondagostino — COMMENTED** [review]
LGTM assuming tests pass

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 1]
```diff
@@ -673,28 +673,30 @@ private KafkaConsumer(ConsumerConfig config, Deserializer<K> keyDeserializer, De
...
+            this.groupId = Optional.ofNullable(groupRebalanceConfig.groupId);
             this.clientId = buildClientId(config.getString(CommonClientConfigs.CLIENT_ID_CONFIG), groupRebalanceConfig);
 
             LogContext logContext;
 
             // If group.instance.id is set, we will append it to the log context.
             if (groupRebalanceConfig.groupInstanceId.isPresent()) {
                 logContext = new LogContext("[Consumer instanceId=" + groupRebalanceConfig.groupInstanceId.get() +
-                        ", clientId=" + clientId + ", groupId=" + groupId + "] ");
+                        ", clientId=" + clientId + ", groupId=" + groupId.orElse("null") + "] ");
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:684]: Perhaps we can improve this. If the groupId is null, perhaps we can leave it out of the log context?

Comment (belugabehr) [reply clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:684]: I admit that 'null' is not flattering, but I was preserving the current behavior.  Would you like something else?  We can improve this is a follow-up JIRA if you'd like to discuss.

**Comment by hachikuji** [comment]
ok to test

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java (hunk 1 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:684]: We can do it separately if you like.

**Comment by hachikuji** [comment]
retest this please

**Review by hachikuji — APPROVED** [review]
LGTM

**Comment by hachikuji** [comment]
retest this please

**Pull request merged**